File: July-LeetCoding-Challenge/22-Binary-Tree-Zigzag-Level-Order-Traversal/Binary-Tree-Zigzag-Level-Order-Traversal.cs
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        var res = new List<IList<int>>();
        if(root == null) return res;
        var level = 0;
        Dfs(root, level, res);
        return res;
    }
    
    private void Dfs(TreeNode node, int level, List<IList<int>> res){
        if(node == null) return;
        if(res.Count <= level) res.Add(new List<int>());
        
        if(level % 2 == 1) res[level].Insert(0, node.val);
        else res[level].Add(node.val);
        
        Dfs(node.left, level+1, res);
        Dfs(node.right, level+1, res);
    }
}
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

File: July-LeetCoding-Challenge/22-Binary-Tree-Zigzag-Level-Order-Traversal/Binary-Tree-Zigzag-Level-Order-Traversal.cpp
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> r;
        queue<TreeNode*> q;
        if(root) q.push(root);
        int h = 1, n;
        while(!q.empty()){
            n = q.size();
            vector<int> ins;
            for(int i=0; i<n; ++i){
                auto node = q.front(); q.pop();                                   
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
                ins.push_back(node->val);
            }
            if(!(h&1))
                reverse(ins.begin(),ins.end());
            r.push_back(ins);                
            ++h;
        }
        return r;
    }
};
